File: Mathematical_Algorithms/Python/exponentiation_by_squaring.py

def exponentiation_by_squaring(base, exp, mod=None):
    res = 1
    base %= mod if mod else base # Đảm bảo base nằm trong phạm vi modulo nếu có

    while exp > 0:
        # Nếu exp là lẻ, nhân base với res
        if exp % 2 == 1:
            res = (res * base) % mod if mod else (res * base)

        # exp trở thành exp/2
        base = (base * base) % mod if mod else (base * base)
        exp //= 2

    return res

File: Mathematical_Algorithms/Java/exponentiation_by_squaring.java
public class exponentiation_by_squaring {

    public static long power(long base, long exp, Long mod) {
        long res = 1;
        if (mod != null) {
            base %= mod;
        }

        while (exp > 0) {
            if (exp % 2 == 1) {
                if (mod != null) {
                    res = (res * base) % mod;
                } else {
                    res = (res * base);
                }
            }
            if (mod != null) {
                base = (base * base) % mod;
            } else {
                base = (base * base);
            }
            exp /= 2;
        }
        return res;
    }

    public static void main(String[] args) {
        System.out.println("2^10 = " + power(2, 10, null));
        System.out.println("3^5 mod 7 = " + power(3, 5, 7L));
        System.out.println("5^13 = " + power(5, 13, null));
    }
}
